File: library/src/main/java/com/lush/lib/adapter/BaseListAdapter.java
package com.lush.lib.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.lush.lib.listener.OnListItemClickListener;
import com.lush.view.holder.BaseViewHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Matt Allen
 */
public abstract class BaseListAdapter<T> extends RecyclerView.Adapter<BaseViewHolder<T>>
{
	private List<T> mItems;
	private OnListItemClickListener<T> mListener;

	public BaseListAdapter(List<T> items, OnListItemClickListener<T> listener)
	{
		setItems(items);
		mListener = listener;
	}

	@Override
	public int getItemCount()
	{
		return mItems != null ? mItems.size() : 0;
	}

	public T getItem(int position)
	{
		return mItems != null ? mItems.get(position) : null;
	}

	public void setItems(List<T> items)
	{
		if (items != null)
		{
			mItems = new ArrayList<>(items);
			notifyDataSetChanged();
		}
	}

	public List<T> getItems()
	{
		if (mItems == null)
		{
			mItems = new ArrayList<>();
		}
		return mItems;
	}

	public void addItem(T item)
	{
		if (mItems == null)
		{
			mItems = new ArrayList<>();
		}
		mItems.add(item);
		notifyItemInserted(mItems.indexOf(item));
	}

	public void removeItem(T item)
	{
		if (mItems != null)
		{
			notifyItemRemoved(mItems.indexOf(item));
			mItems.remove(item);
		}
	}

	public T removeItem(int position)
	{
		if (mItems != null)
		{
			T item = mItems.remove(position);
			notifyItemRemoved(position);
			return item;
		}
		return null;
	}

	public void addItem(int position, T item)
	{
		getItems().add(position, item);
		notifyItemInserted(position);
	}

	public void moveItem(int fromPosition, int toPosition)
	{
		if (mItems != null)
		{
			T item = mItems.remove(fromPosition);
			mItems.add(toPosition, item);
			notifyItemMoved(fromPosition, toPosition);
		}
	}

	protected OnListItemClickListener<T> getListener()
	{
		return mListener;
	}

	protected int getPosition(T item)
	{
		if (mItems != null && mItems.size() > 0)
		{
			return mItems.indexOf(item);
		}
		return -1;
	}

	public void setItemsWithAnimation(List<T> items)
	{
		applyAndAnimateRemovals(items);
		applyAndAnimateAdditions(items);
		applyAndAnimateMovedItems(items);
	}

	private void applyAndAnimateRemovals(List<T> newItems)
	{
		for (int i = getItems().size() - 1; i >= 0; i--)
		{
			T item = getItems().get(i);
			if (!newItems.contains(item))
			{
				removeItem(i);
			}
		}
	}

	private void applyAndAnimateAdditions(List<T> newItems)
	{
		for (int i = 0, count = newItems.size(); i < count; i++)
		{
			T item = newItems.get(i);
			if (!getItems().contains(item))
			{
				addItem(i, item);
			}
		}
	}

	private void applyAndAnimateMovedItems(List<T> newItems)
	{
		for (int toPosition = newItems.size() - 1; toPosition >= 0; toPosition--)
		{
			T model = newItems.get(toPosition);
			final int fromPosition = getItems().indexOf(model);
			if (fromPosition >= 0 && fromPosition != toPosition)
			{
				moveItem(fromPosition, toPosition);
			}
		}
	}

	public void setListener(OnListItemClickListener<T> listener)
	{
		mListener = listener;
	}

	@Override
	public void onBindViewHolder(BaseViewHolder<T> holder, int position)
	{
		final T item = getItem(position);
		holder.bind(item);
		if (mListener != null)
		{
			holder.setOnClickListener(new View.OnClickListener()
			{
				@Override
				public void onClick(View v)
				{
					mListener.onItemClick(item, v);
				}
			});
		}
	}
}

File: views/src/main/java/com/lush/lib/adapter/BaseListAdapter.kt
package com.lush.lib.adapter

import android.support.v7.widget.RecyclerView
import android.view.View
import com.lush.lib.listener.OnListItemClickListener
import com.lush.view.holder.BaseViewHolder

abstract class BaseListAdapter<T>(
		proposedItems: List<T> = ArrayList(),
		protected val listener: OnListItemClickListener<T>? = null,
		private val colourAlternateItems: Boolean
): RecyclerView.Adapter<BaseViewHolder<T>>()
{
	private val items: ArrayList<T> = ArrayList(proposedItems.filter { it != null })

	constructor(proposedItems: List<T> = ArrayList(), listener: OnListItemClickListener<T>? = null): this(proposedItems, listener, false)

	override fun onBindViewHolder(holder: BaseViewHolder<T>, position: Int)
	{
		val item = getItem(position)
		holder.bind(item)
		if (listener != null)
		{
			holder.setOnClickListener(View.OnClickListener { listener.onItemClick(item, it) })
		}
		if (colourAlternateItems)
		{
			holder.setPositionInList(position)
		}
	}

	override fun onViewRecycled(holder: BaseViewHolder<T>)
	{
		super.onViewRecycled(holder)
		holder.recycle()
	}

	override fun getItemCount(): Int = items.size

	fun addItem(item: T)
	{
		items.add(item)
		notifyItemInserted(items.indexOf(item))
	}

	fun removeItem(item: T)
	{
		notifyItemRemoved(items.indexOf(item))
		items.remove(item)
	}

	fun removeItem(position: Int): T
	{
		val item = items.removeAt(position)
		notifyItemRemoved(position)
		return item
	}

	fun addItem(position: Int, item: T)
	{
		items.add(position, item)
		notifyItemInserted(position)
	}

	fun moveItem(fromPosition: Int, toPosition: Int)
	{
		val item = items.removeAt(fromPosition)
		if (toPosition >= items.size)
		{
			items.add(item)
		}
		else
		{
			items.add(toPosition, item)
		}
		notifyItemMoved(fromPosition, toPosition)
	}

	fun setItemsWithAnimation(newItems: List<T>)
	{
		val itemsToSet = newItems.filter { it != null }
		applyAndAnimateRemovals(itemsToSet)
		applyAndAnimateAdditions(itemsToSet)
		applyAndAnimateMovedItems(itemsToSet)
		notifyItemRangeChanged(0, itemsToSet.size)
	}

	fun setItems(newItems: List<T>)
	{
		items.clear()
		items.addAll(newItems.filter { it != null })
		notifyDataSetChanged()
	}

	fun getItems(): List<T> = ArrayList(items)
	fun getItem(position: Int): T = items[position]

	protected fun getPosition(item: T): Int
	{
		return if (items.size > 0)
		{
			items.indexOf(item)
		}
		else -1
	}

	private fun applyAndAnimateRemovals(newItems: List<T>)
	{
		for (i in items.size - 1 downTo 0)
		{
			val item = items[i]
			if (!newItems.contains(item))
			{
				removeItem(i)
			}
		}
	}

	private fun applyAndAnimateAdditions(newItems: List<T>)
	{
		var i = 0
		val count = newItems.size
		while (i < count)
		{
			val item = newItems[i]
			if (!items.contains(item))
			{
				addItem(i, item)
			}
			i++
		}
	}

	private fun applyAndAnimateMovedItems(newItems: List<T>)
	{
		for (toPosition in newItems.indices.reversed())
		{
			val model = newItems[toPosition]
			val fromPosition = items.indexOf(model)
			if (fromPosition >= 0 && fromPosition != toPosition)
			{
				moveItem(fromPosition, toPosition)
			}
		}
	}
}